text/0000-refs.md

```diff
@@ -0,0 +1,209 @@
+
+### Multiple templates
+
+In the case of multiple templates, the same rules as for `this.template.querySelector()` or `this.querySelector()` apply. Once the new template is rendered, `this.refs` refers to elements defined in the new template:
+
+```js
+import a from './a.html';
+import b from './b.html';
+
+export default class extends LightningElement {
+  count = 0;
+  
+  render() {
+    return this.count % 2 === 0 ? a : b;
+  }
+  
+  renderedCallback() {
+    console.log(this.refs);
+  }
+  
+  increment () {
+    this.count++;
+  }
+}
+```
+
+```js
+// Logs `this.refs` for a.html
+const cmp = createElement('x-component', { is: Component });
+
+// Logs `this.refs` for b.html
+cmp.increment();
+
+// Logs `this.refs` for a.html
+cmp.increment();
+```
+
+### Duplicate `lwc:ref`s
+
+If duplicate `lwc:ref`s are defined in the same template:
+
+```html
+<template>
+    <div lwc:ref="foo"></div>
+    <div lwc:ref="foo"></div>
+</template>
+```
+
+... then the second one "wins" (in terms of depth-first tree order). The reason for allowing duplicates is explained below.
+
+### Conditional elements
+
+In some cases, it may be useful to have elements that are conditionally defined based on `<template if:true>`. For example:
+
+```html
+<template>
+    <template if:true={darkMode}>
+        <div lwc:ref="foo"></div>      
+    </template>
+    <template if:false={darkMode}>
+        <div lwc:ref="foo"></div>
+    </template>
+</template>
+```
+
+In this case, it's convenient to be able to call `this.refs.foo` and to not care whether the element is inside the first `<template>` or the second.
+
+Because the template compiler cannot know which conditions will cause an element to exist or not, we cannot support this use case while also disallowing duplicates. Therefore, the simplest solution for duplicates is to define whichever element is last (in terms of depth-first tree order) to be the one that "wins."
+
+Here is how other frameworks do it:
+
+- React: [duplicates allowed, last one wins](https://codepen.io/nolanlawson-the-selector/pen/dyVvZwZ)
+- Vue: [duplicates allowed, last one wins](https://codepen.io/nolanlawson-the-selector/pen/jOGBadg)
+- Svelte: [duplicates allowed, last one wins](https://svelte.dev/repl/db716cce22b84cc79e8612417ad7a804?version=3.44.3)
+
+## Drawbacks
+
+This feature may cause some confusion over what is the "best" way to access DOM elements in a component – `querySelector` (and other DOM APIs) or refs.
+
+Also, refs are non-standard, whereas `querySelector` and friends are standard.
+
+## Alternatives
+
+The impact of _not_ doing this is that it's much trickier to find the right elements, in light DOM in particular.
+
+## Adoption strategy
+
+Technically this is a breaking change, because developers may have already defined a property named `refs` in their component. We will need to do some analysis of potential breakage and find workarounds (e.g. ESLint rules).

```

Comment:
I don't think it is a breaking change if the value is configurable/writable. This will be defined in the `LightningElement.prototype`, meaning if anyone is declaring that, or attempting to redefining it, should be able to do so, overruling the default behavior, meaning it will not interfere with whatever they are doing today.

Reply:
The only observable change would be if the component access `this.refs` without defining it first. But I am not much worried about this.

```js
class Example extends LightningElement {
  renderedCallback() {
    if (this.refs) {  // <-- Will always return true
        this.refs.div.textContent += '[updated]!';
    } else {
        this.refs = { 
          div: this.template.querySelector('div') 
        };
    }
}
```

Reply:
Ah okay, this is a great point. I updated the document.